src/main/scala/edu/ie3/osmogrid/io/input/InputDataProvider.scala

```diff
@@ -6,73 +6,145 @@
-  final case class RepOsm(runId: UUID, osmModel: OsmoGridModel) extends Response
-  final case class InvalidOsmRequest(reqRunId: UUID, actualRunId: UUID)
+  final case class RepOsm(runId: UUID, osmModel: OsmoGridModel)
+      extends Response
+      with InputDataEvent
+  final case class OsmReadFailed(reason: Throwable)
+      extends Response
+      with InputDataEvent
+  final case class RepAssetTypes(assetInformation: AssetInformation)
       extends Response
-  final case class OsmReadFailed(reason: Throwable) extends Response
-  final case class RepAssetTypes(
-      assetInformation: AssetInformation
-  ) extends Response
 
   final case class AssetInformation(
       lineTypes: Seq[LineTypeInput],
       transformerTypes: Seq[Transformer2WTypeInput]
   )
 
-  def apply(cfg: OsmoGridConfig.Input): Behavior[Request] =
-    Behaviors
-      .receive[Request] {
-        case (ctx, ReqOsm(runId, replyTo, filter)) =>
-          ctx.log.warn("Reading of data not yet implemented.")
-          Behaviors.same
-        case (ctx, ReqAssetTypes(_, _)) =>
+  // internal api
+  sealed trait InputDataEvent
+
+  // actor data
+  private final case class ProviderData(
+      runId: UUID,
+      ctx: ActorContext[InputDataEvent],
+      buffer: StashBuffer[InputDataEvent],
+      osmSource: OsmSource
+  )
+
+  def apply(
+      runId: UUID,
+      osmConfig: OsmoGridConfig.Input
+  ): Behavior[InputDataEvent] = {
+    Behaviors.withStash[InputDataEvent](100) { buffer =>
+      Behaviors.setup[InputDataEvent] { ctx =>
+        idle(
+          ProviderData(runId, ctx, buffer, OsmSource.apply(osmConfig.osm, ctx))
+        )
+      }
+    }
+  }
+
+  private def idle(providerData: ProviderData): Behavior[InputDataEvent] =
+    Behaviors.receive[InputDataEvent] { case (ctx, msg) =>
+      msg match {
+        case ReqOsm(replyTo, filter) =>
+          ctx.pipeToSelf(
+            providerData.osmSource.read(filter)
+          ) {
+            case Success(osmoGridModel: OsmoGridModel) =>
+              RepOsm(providerData.runId, osmoGridModel)
+            case Failure(exception) =>
+              ctx.log.error(
+                s"Error while reading osm data: $exception"
+              )
+              OsmReadFailed(exception)
+          }
+          readOsmData(providerData, replyTo)
+        case ReqAssetTypes(_, _) =>
           ctx.log.info("Got request to provide asset types. But do nothing.")
           Behaviors.same
-        case (ctx, Terminate) =>
-          ctx.log.info("Stopping input data provider")
-          stopBehavior
+        case Terminate =>
+          ctx.log.info("Stopping input data provider ...")
+          cleanUp(providerData)
+          Behaviors.stopped
+        case invalid: (OsmReadFailed | RepOsm) =>
+          ctx.log.error(
+            s"Received unexpected message '$invalid' in state Idle! Ignoring!"
+          )
+          Behaviors.same
       }
-      .receiveSignal { case (ctx, PostStop) => postStopCleanUp(ctx.log) }
```

Comment:
What is the reason, that this line is gone? I intentionally inserted it to catch the case, where the actor is terminated via `ActorContext#stop`

Reply:
Don't know why it was gone, re-added